PSP ephemeral volume validation

pohly opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

As pointed out in https://github.com/kubernetes/kubernetes/commit/c05c8e915b0c30d574405ed63e48a52801a1f75b#r46791570 the introduction of the "generic" volume type did not consider that this is an implicit API change for PodSecurityPolicy which must take the feature gate into account.

This gets fixed and one more test case for applying the PSP gets added.

**Special notes for your reviewer**:

I think the validation itself (pkg/apis/policy/validation/validation.go) does not need to check the feature gate because the new type will already have been dropped depending on the feature gate and the old object (https://github.com/kubernetes/kubernetes/pull/80568#discussion_r314817559).

**Does this PR introduce a user-facing change?**:
```release-note
PodSecurityPolicy only stores "generic" as allowed volume type if the GenericEphemeralVolume feature gate is enabled
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes
```

pohly committed:
security: another test case for generic ephemeral inline volumes

When the PSP contains some other volume types, generic ephemeral
inline volumes must be rejected.

pohly committed:
PSP: conditional support for generic volume type

When introducing the new "generic" volume type for generic ephemeral
inline volumes, the storage policy for PodSecurityPolicy objects
should have been extended so that this new type is valid only
if the generic ephemeral volume feature is enabled or an
existing object already has it.

Adding the new type to the internal API was also missed.

pohly commented:
/sig security

pohly commented:
/retest

pohly commented:
/label api-review

liggitt commented on pkg/api/podsecuritypolicy/util.go:
```diff
@@ -31,6 +31,16 @@ func DropDisabledFields(pspSpec, oldPSPSpec *policy.PodSecurityPolicySpec) {
 	}
+	var volumes []policy.FSType
+	for _, volume := range pspSpec.Volumes {
+		if volume != policy.Ephemeral ||
```
A new enum value for an existing field has to be handled differently than a completely new field. Modifying a user-specified list by dropping a value from it is problematic. 

The goal is to make a current server with the feature disabled behave as much as possible like a server that is unaware of the feature, with the exception of safely handling already persisted data that uses the field.

I would suggest doing the following:

1. in `pkg/apis/policy/validation/validation.go`, add a PodSecurityPolicyValidationOptions type containing a single `AllowEphemeralVolumeType bool` field
2. pass that to ValidatePodSecurityPolicy, ValidatePodSecurityPolicyUpdate, ValidatePodSecurityPolicySpec, and validatePodSecurityPolicyVolumes
3. in validatePodSecurityPolicyVolumes, only allow the volumes list to contain `policy.Ephemeral` if `opts.AllowEphemeralVolumeType` is true
4. in `pkg/registry/policy/podsecuritypolicy/strategy.go`, compute the validation options to pass to ValidatePodSecurityPolicy and ValidatePodSecurityPolicyUpdate based on the new and old object and the feature enablement. https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/resourcequota/strategy.go#L104 is a good example to follow

liggitt replied:
The combinations to test are similar, but instead of dropping the field, we should reject in validation:

with feature disabled
* create psp with "ephemeral" value → fail validation
* update existing psp object without "ephemeral" value to new object with "ephemeral" value → fail validation
* update existing psp object with "ephemeral" value to new object with "ephemeral" value → allowed
* update existing psp object with "ephemeral" value to new object without "ephemeral" value → allowed

feature enabled:
* create psp with "ephemeral" value → allowed
* update existing psp object without "ephemeral" value to new object with "ephemeral" value → allowed
* update existing psp object with "ephemeral" value to new object with "ephemeral" value → allowed
* update existing psp object with "ephemeral" value to new object without "ephemeral" value → allowed

liggitt replied:
opened https://github.com/kubernetes/community/pull/5607/files to capture this guidance for adding new permitted enum values

pohly replied:
> The goal is to make a current server with the feature disabled behave as much as possible like a server that is unaware of the feature

Makes sense. What I hadn't considered is that such a server treats unknown fields and unknown enums differently: it complains about the unknown enum and silently ignores the unknown field, right? I've always wondered whether the latter was intentional or a side-effect of using the Go JSON parser.

I've made the suggested changes, thanks for the detailed pointers. I wasn't sure where to do the unit testing. Definitely in `pkg/registry/policy/podsecuritypolicy`, but are tests then also needed in `pkg/apis/policy/validation`? For the sake of completeness yes, but in practice they are almost identical and cover the same code paths.

I now have them in both places. I can remove them from `pkg/apis/policy/validation` if those are seen as redundant.

liggitt replied:
if I were writing the unit tests from scratch, I'd probably split the PodSecurityPolicyValidationOptions computating into its own function in `strategy.go` and unit test that in the strategy unit tests, and have the validation unit tests focus on the combinations of the object being validated and the AllowEphemeralVolumeType field

since you already have the unit tests written, I'd probably leave them as-is.

pohly committed:
PSP: validation errors for generic volume type

It's not enough to silently drop the volume type if the feature is
disabled. Instead, the policy should fail validation, just as it would
have if the API server didn't know about the feature at all.

liggitt commented:
/lgtm
/approve
/triage accepted
/priority important-soon
